MAINT: Update ndarrayobject.h `__cplusplus` block.

lerbuke opened the pull request:
Complete (in the same way than done in line 8 to 14) the __cplusplus check to avoid the following warning:
C:/numpy/core/include/numpy/ndarrayobject.h:243: Excess closing brace in C++ code (or abuse of the C++ preprocessor)

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html#writing-the-commit-message
-->

lerbuke committed:
Update ndarrayobject.h

Complete (in the same way than done in line 8 to 14) the __cplusplus check to avoid the following warning:
C:/numpy/core/include/numpy/ndarrayobject.h:243: Excess closing brace in C++ code (or abuse of the C++ preprocessor)

tylerjereddy commented on numpy/core/include/numpy/ndarrayobject.h:
```diff
@@ -284,7 +284,9 @@ PyArray_XDECREF_ERR(PyArrayObject *arr)
-}
+#define CONFUSE_EMACS2 }
+CONFUSE_EMACS2
+#undef CONFUSE_EMACS2
```
Why is this block shorter than the other one you've mentioned in the same file, which looks like this:

```C
#ifdef __cplusplus
#define CONFUSE_EMACS {
#define CONFUSE_EMACS2 }
extern "C" CONFUSE_EMACS
#undef CONFUSE_EMACS
#undef CONFUSE_EMACS2
/* ... otherwise a semi-smart identer (like emacs) tries to indent
       everything when you're typing */
#endif
```

lerbuke commented:
Hello,

I think the original author maybe done a mistake because CONFUSE_EMACS2 is defined and then undefined without being used in the __cplusplus scope, which is normal as we need only the CONFUSE_EMACS at this location the file.

The extern ”C” declaration makes function names in C++ have a 'C' linkage, and if you need to declare several function names, you put them into {}

// One function only, one line declaration without need of {}
extern ”C” void foo(double); 

// Several functions, we need to put them into {}
extern ”C” 
{
	void foo1(double); 
	int foo2();
	…
	void start();
}

Regards,
Levent

> On 11 Oct 2018, at 20:02, Tyler Reddy <notifications@github.com> wrote:
> 
> @tylerjereddy commented on this pull request.
> 
> In numpy/core/include/numpy/ndarrayobject.h <https://github.com/numpy/numpy/pull/12145#discussion_r224547099>:
> 
> > @@ -284,7 +284,9 @@ PyArray_XDECREF_ERR(PyArrayObject *arr)
>  
>  
>  #ifdef __cplusplus
> -}
> +#define CONFUSE_EMACS2 }
> +CONFUSE_EMACS2
> +#undef CONFUSE_EMACS2
> Why is this block shorter than the other one you've mentioned in the same file, which looks like this:
> 
> #ifdef __cplusplus
> #define CONFUSE_EMACS {
> #define CONFUSE_EMACS2 }
> extern "C" CONFUSE_EMACS
> #undef CONFUSE_EMACS
> #undef CONFUSE_EMACS2
> /* ... otherwise a semi-smart identer (like emacs) tries to indent
>        everything when you're typing */
> #endif
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/numpy/numpy/pull/12145#pullrequestreview-163942811>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AlFLKMOEe9tIWoEpZP_QCYxmw2kZBkh2ks5uj4ehgaJpZM4XXhMH>.
>

mattip commented:
It would be nice to simply remove both blocks, if emacs has fixed its smart indenting problems. Any emacs users care to comment?

eric-wieser commented:
What's so special about `ndarrayobject.h` that we care about emacs in it, but no other headers?

lerbuke committed:
Removed all useless definitions in __cplusplus blocks.

Followed @mattip remark which I think is the best solution:

https://github.com/numpy/numpy/pull/12145#issuecomment-429723159

mattip commented:
Thanks @lerbuke
